core/src/main/java/org/infinispan/stream/impl/local/EntryStreamSupplier.java

```diff
@@ -44,7 +45,9 @@ public EntryStreamSupplier(Cache<K, V> cache, ConsistentHash hash, Supplier<Stre
       }
       if (segmentsToFilter != null && hash != null) {
          log.tracef("Applying segment filter %s", segmentsToFilter);
-         stream = stream.filter(k -> segmentsToFilter.contains(hash.getSegment(k.getKey())));
+         BitSet bitSet = new BitSet(hash.getNumSegments());
```

Comment:
BitSet seems like a good idea. I think it's worth changing StreamSupplier.buildStream to take a BitSet for segmentsToFilter.  I'd also add a check for log.isTraceEnabled() to avoid the toString().